pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go

```diff
@@ -222,136 +306,220 @@ func (m *rangefeedMuxer) startNodeMuxRangeFeed(
 	defer restore()
 
 	if log.V(1) {
-		log.Info(ctx, "Establishing MuxRangeFeed")
+		log.Infof(ctx, "Establishing MuxRangeFeed to node %d", nodeID)
 		start := timeutil.Now()
 		defer func() {
-			log.Infof(ctx, "MuxRangeFeed terminating after %s", timeutil.Since(start))
+			log.Infof(ctx, "MuxRangeFeed to node %d terminating after %s with err=%v",
+				nodeID, timeutil.Since(start), retErr)
 		}()
 	}
 
-	doneCtx := makeTerminationContext()
-	ms.doneCtx = &doneCtx
 	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	mux, err := client.MuxRangeFeed(ctx)
+	if err != nil {
+		conn.Set(connOrError{err: err})
+		return nil //nolint:returnerrcheck
+	}
 
-	ms.cancel = func() {
-		cancel()
-		doneCtx.close(context.Canceled)
+	nodeConn := &muxNodeConn{
+		sender:   &lockedRangeFeedRequestSender{rangeFeedRequestSender: mux},
+		receiver: mux,
 	}
-	defer ms.cancel()
-
-	// NB: it is important that this Go routine never returns an error. Errors
-	// should be propagated to the caller either via initCtx.err, or doneCtx.err.
-	// We do this to make sure that this error does not kill entire context group.
-	// We want the caller (singleRangeFeed) to decide if this error is retry-able.
-	var err error
-	ms.client, err = client.MuxRangeFeed(ctx)
-	ms.initCtx.close(err)
-
-	if err == nil {
-		doneCtx.close(m.receiveEventsFromNode(ctx, ms))
+	conn.Set(connOrError{conn: nodeConn})
+
+	stuckWatcher := newStuckRangeFeedCanceler(cancel, defaultStuckRangeThreshold(m.ds.st))
+	defer stuckWatcher.stop()
+
+	if recvErr := m.receiveEventsFromNode(ctx, nodeID, stuckWatcher, nodeConn); recvErr != nil {
+		// Clear out this client, and restart all streams on this node.
+		// Note: there is a race here where we may delete this muxClient, while
+		// another go routine loaded it.  That's fine, since we would not
+		// be able to send new request on this stream anymore, and we'll retry
+		// against another node.
+		m.muxClients.Delete(int64(nodeID))
+
+		if recvErr == io.EOF {
+			recvErr = nil
+		}
+
+		if _, err := handleRangefeedError(ctx, recvErr); err != nil {
```

Comment:
Shouldn't we properly handle cache eviction here, to avoid unnecessary cache invalidation on unrelated errors?